packages/protobuf/src/extension-accessor.ts

```diff
@@ -0,0 +1,125 @@
+// Copyright 2021-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import type { Message } from "./message.js";
+import type { BinaryReadOptions, BinaryWriteOptions } from "./binary-format.js";
+import { assert } from "./private/assert.js";
+import {
+  createExtensionContainer,
+  filterUnknownFields,
+} from "./private/extensions.js";
+import type { Extension } from "./extension.js";
+
+/**
+ * Retrieve an extension value from a message.
+ *
+ * The function never returns undefined. Use hasExtension() to check whether an
+ * extension is set. If the extension is not set, this function returns the
+ * default value (if one was specified in the protobuf source), or the zero value
+ * (for example `0` for numeric types, `[]` for repeated extension fields, and
+ * an empty message instance for message fields).
+ */
+export function getExtension<E extends Message<E>, V>(
+  message: E,
+  extension: Extension<E, V>,
+  options?: Partial<BinaryReadOptions>,
+): V {
+  const opt = extension.runtime.bin.makeReadOptions(options);
+  assert(!extension.field.oneof); // oneof is not allowed for extensions
+  assert(extension.field.kind != "map"); // maps are not allowed for extensions
```

Comment:
Is it possible to also assert here that the type name of the message matches the extension's extendee? I'm thinking about possible dynamic message cases, where a concrete type `E` of the message is not sufficient to guarantee that the extension correctly applies to it.

Reply:
Good call, added an assertion for the mutating functions, but left the behavior of `hasExtension` as is: https://github.com/bufbuild/protobuf-es/pull/666/commits/c1d5b71f7e7a885ed3dcfb5ec469d00c7eaaead7